### DISCUSSION TITLE
Is there a simpler way to filter for a single document, using more than one parameter?

### DISCUSSION THREAD

**Discussion by NetOpWibby** [edgedb/edgedb-js#527] [Q&A] [answered]
```ts
  const doesDocumentExist = e.select(e.Registrant, registrant => ({
    ...e.Registrant["*"],
    registrar: registrant.registrar["*"],
    // TODO
    // : https://github.com/edgedb/edgedb-js/issues/347
    filter_single: e.op(
      e.op(registrant.email, "=", query.email),
      "and",
      e.op(
        e.op(registrant.registrar.id, "=", e.uuid(query.registrar)),
        "and",
        e.op(registrant.name, "=", query.name)
      )
    )
  }));
```

This is what I have but I don't like it.

What this code is doing: given an object with keys email, registrar, and/or name, bring back the document that matches those defined params.

Not sure where the linked issue is on the internal list of importance but I'd love to have a better solution in the meantime. https://github.com/edgedb/edgedb-js/issues/347

**Comment by NetOpWibby** [comment]
```
No function overload found for operator '=' with args: [object Object], undefined
```

This is the error I get with my function.

**Comment by petersamokhin** [comment, marked as answer]
A dirty workaround but will help:


```ts
export function combineOp(
  operator: string,
  ...inputs: $expr_Operator<any, any>[]
): $expr_Operator<any, any> | undefined {
  if (inputs.length === 0) {
    return undefined
  }

  if (inputs.length === 1) {
    return inputs[0]
  }

  return e.op(
    inputs[0] as $expr_Operator<any, any>,
    operator as never,
    combineOp(operator, ...inputs.slice(1)) as $expr_Operator<any, any>,
  )
}

export function andOp(
  ...inputs: $expr_Operator<any, any>[]
): $expr_Operator<any, any> | undefined {
  return combineOp('and', ...inputs)
}

export function orOp(
  ...inputs: $expr_Operator<any, any>[]
): $expr_Operator<any, any> | undefined {
  return combineOp('or', ...inputs)
}
```

then use it like this:
```ts
e.select(e.Foo, () => ({
  // ...
  filter: andOp(e.op(foo, '=', bar), e.op(bar, '!=', baz)),
})
```

or providing an array, or a single element, etc.

**Reply by NetOpWibby** [reply]
Can I use strict equality with this?

**Reply by petersamokhin** [reply]
@NetOpWibby the purpose of that workaround is to combine multiple `e.op` — and `e.op` itself can include anything. If I correctly understood you

**Reply by NetOpWibby** [reply]
@petersamokhin I was talking about `!=` in the example above. Is it interchangeable with `!==`?

**Reply by petersamokhin** [reply]
it can be whatever e.op supports, it is directly used in the result edgeql query which doesn't use !==, but uses != etc

**Reply by NetOpWibby** [reply]
Oh sweet, thank you! I'll try this out today.

**Reply by NetOpWibby** [reply]
I'm late but this works wonderfully. Thanks again @petersamokhin!

**Reply by NetOpWibby** [reply]
Half a year later with an update.

I finally had reason to use `orOp` and I ran into an error (`Invalid value for type std::uuid: undefined`). To get around it, I just did this:

```js
    ...(query.extension ? {
      filter: e.op(document.extension.id, "=", e.uuid(query.extension))
    } : {}),
    ...(query.owner ? {
      filter: e.op(document.owner.id, "=", e.uuid(query.owner))
    } : {}),
```

**Reply by NetOpWibby** [reply]
I am quite thankful for how helpful past me is to future me.
